// geo-map/lib/result/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:no-any
var Types = require("../types");
function createResult() {
    return {
        type: Types.ResultType.Queued,
        startedAt: new Date()
    };
}
exports.createResult = createResult;
function createSuccess(payload) {
    return {
        type: Types.ResultType.Success,
        payload: payload
    };
}
exports.createSuccess = createSuccess;
exports.createFailure = function (error) {
    return {
        type: Types.ResultType.Failure,
        error: error
    };
};
function from(data) {
    if (data.type === Types.ResultType.Failure) {
        return exports.createFailure(data.error);
    }
    if (data.type === Types.ResultType.Success) {
        return createSuccess(data.payload);
    }
    return createResult();
}
exports.from = from;
exports.toFailure = function (result, error) {
    var failure = result;
    delete failure.startedAt;
    delete failure.payload;
    failure.type = Types.ResultType.Failure;
    failure.error = error;
    return failure;
};
exports.toSuccess = function (result, payload) {
    var success = result;
    delete success.startedAt;
    delete success.error;
    success.type = Types.ResultType.Success;
    success.payload = payload;
    return success;
};
exports.toQueued = function (result) {
    var queued = result;
    delete queued.payload;
    delete queued.error;
    queued.type = Types.ResultType.Queued;
    queued.startedAt = new Date();
    return queued;
};
