// geo-map/esm/types/geo-map.js
export var GeoMapProvider;
(function (GeoMapProvider) {
    GeoMapProvider["Google"] = "Google";
    GeoMapProvider["Here"] = "Here";
    GeoMapProvider["Custom"] = "Custom";
})(GeoMapProvider || (GeoMapProvider = {}));
export var GoogleMapAuthType;
(function (GoogleMapAuthType) {
    GoogleMapAuthType["ApiKey"] = "ApiKey";
    GoogleMapAuthType["ClientId"] = "ClientId";
})(GoogleMapAuthType || (GoogleMapAuthType = {}));
export var PlaceType;
(function (PlaceType) {
    PlaceType[PlaceType["Unknown"] = 0] = "Unknown";
    PlaceType[PlaceType["Removed"] = 1] = "Removed";
    PlaceType[PlaceType["Car"] = 2] = "Car";
    PlaceType[PlaceType["Favorit"] = 4] = "Favorit";
    PlaceType[PlaceType["LastUsed"] = 8] = "LastUsed";
})(PlaceType || (PlaceType = {}));
export var GeoMapType;
(function (GeoMapType) {
    GeoMapType["Roadmap"] = "Roadmap";
    GeoMapType["Hybrid"] = "Hybrid";
    GeoMapType["Unknown"] = "Unknown";
})(GeoMapType || (GeoMapType = {}));
export var GeoLayer;
(function (GeoLayer) {
    GeoLayer["Transit"] = "Transit";
    GeoLayer["Traffic"] = "Traffic";
    GeoLayer["None"] = "None";
})(GeoLayer || (GeoLayer = {}));
export var GeoMarkerOrientation;
(function (GeoMarkerOrientation) {
    GeoMarkerOrientation[GeoMarkerOrientation["Start"] = 0] = "Start";
    GeoMarkerOrientation[GeoMarkerOrientation["Middle"] = 1] = "Middle";
    GeoMarkerOrientation[GeoMarkerOrientation["End"] = 2] = "End";
})(GeoMarkerOrientation || (GeoMarkerOrientation = {}));
export var HerePixelDensity;
(function (HerePixelDensity) {
    HerePixelDensity[HerePixelDensity["Default"] = 72] = "Default";
    HerePixelDensity[HerePixelDensity["HighRes"] = 320] = "HighRes";
    HerePixelDensity[HerePixelDensity["UltraHighRes"] = 500] = "UltraHighRes";
})(HerePixelDensity || (HerePixelDensity = {}));
export var HereLanguage;
(function (HereLanguage) {
    HereLanguage["ar"] = "ara";
    HereLanguage["eu"] = "baq";
    HereLanguage["ca"] = "cat";
    HereLanguage["zh"] = "chi";
    HereLanguage["cs"] = "cze";
    HereLanguage["da"] = "dan";
    HereLanguage["nl"] = "dut";
    HereLanguage["en"] = "eng";
    HereLanguage["fi"] = "fin";
    HereLanguage["fr"] = "fre";
    HereLanguage["de"] = "ger";
    HereLanguage["ga"] = "gle";
    HereLanguage["el"] = "gre";
    HereLanguage["he"] = "heb";
    HereLanguage["hi"] = "hin";
    HereLanguage["id"] = "ind";
    HereLanguage["it"] = "ita";
    HereLanguage["no"] = "nor";
    HereLanguage["fa"] = "per";
    HereLanguage["pl"] = "pol";
    HereLanguage["pt"] = "por";
    HereLanguage["ru"] = "rus";
    HereLanguage["si"] = "sin";
    HereLanguage["es"] = "spa";
    HereLanguage["sw"] = "swe";
    HereLanguage["th"] = "tha";
    HereLanguage["tr"] = "tur";
    HereLanguage["uk"] = "ukr";
    HereLanguage["ur"] = "urd";
    HereLanguage["vi"] = "vie";
    HereLanguage["cy"] = "wel";
    HereLanguage["Multiple"] = "mul";
})(HereLanguage || (HereLanguage = {}));
export var GeoMapPhase;
(function (GeoMapPhase) {
    GeoMapPhase["Pristine"] = "pristine";
    GeoMapPhase["Loading"] = "loading";
    GeoMapPhase["Loaded"] = "loaded";
    GeoMapPhase["Mounting"] = "mounting";
    GeoMapPhase["Mounted"] = "mounted";
    GeoMapPhase["Layouting"] = "layouting";
    GeoMapPhase["Layouted"] = "layouted";
})(GeoMapPhase || (GeoMapPhase = {}));
export var GeoEvent;
(function (GeoEvent) {
    GeoEvent["Click"] = "click";
    GeoEvent["Changed"] = "changed";
    GeoEvent["Loaded"] = "tilesloaded";
})(GeoEvent || (GeoEvent = {}));
