// geo-map/esm/geo-marker-google.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import * as Result from './result';
import * as Types from './types';
var GeoMarkerGoogle = /** @class */ (function () {
    function GeoMarkerGoogle(config, context) {
        this.anchor = {
            vertical: Types.GeoMarkerOrientation.Middle,
            horizontal: Types.GeoMarkerOrientation.Middle
        };
        this.implementation = context.mapImplementation;
        this.iconMarkup = config.icon;
        if (config.anchor) {
            this.anchor = config.anchor;
        }
        var iconAnchor = getAnchor(this.anchor, config.icon);
        this.marker = new this.implementation.api.Marker({
            position: config.position,
            map: this.implementation.map,
            icon: {
                anchor: iconAnchor
                    ? new this.implementation.api.Point(iconAnchor.x, iconAnchor.y)
                    : undefined,
                url: "data:image/svg+xml;utf-8," + encodeURIComponent(config.icon)
            }
        });
        this.implementation.markers.push(this);
        this.implementation.fire(Types.GeoEvent.Changed);
    }
    GeoMarkerGoogle.create = function (config, context) {
        return new GeoMarkerGoogle(config, context);
    };
    GeoMarkerGoogle.prototype.getIcon = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.iconMarkup];
            });
        });
    };
    GeoMarkerGoogle.prototype.setIcon = function (icon) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.iconMarkup = icon;
                this.marker.setIcon({
                    anchor: new this.implementation.api.Point(16, 16),
                    url: "data:image/svg+xml;utf-8," + encodeURIComponent(icon)
                });
                return [2 /*return*/, Result.createSuccess()];
            });
        });
    };
    GeoMarkerGoogle.prototype.setPosition = function (position) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.marker.setPosition(position);
                return [2 /*return*/];
            });
        });
    };
    GeoMarkerGoogle.prototype.getPosition = function () {
        return __awaiter(this, void 0, void 0, function () {
            var position;
            return __generator(this, function (_a) {
                position = this.marker.getPosition();
                return [2 /*return*/, {
                        lat: position.lat(),
                        lng: position.lng()
                    }];
            });
        });
    };
    GeoMarkerGoogle.prototype.remove = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.marker.setMap(null);
                this.implementation.markers.splice(this.implementation.markers.indexOf(this));
                this.implementation.fire(Types.GeoEvent.Changed);
                return [2 /*return*/];
            });
        });
    };
    return GeoMarkerGoogle;
}());
export { GeoMarkerGoogle };
function getOrientationRatio(orientation) {
    switch (orientation) {
        case Types.GeoMarkerOrientation.Start:
            return 0;
        case Types.GeoMarkerOrientation.End:
            return 1;
        case Types.GeoMarkerOrientation.Middle:
        default:
            return 0.5;
    }
}
function getAnchor(anchor, icon) {
    if (typeof window === 'undefined') {
        return;
    }
    var wRatio = getOrientationRatio(anchor.horizontal);
    var hRatio = getOrientationRatio(anchor.vertical);
    var parser = new DOMParser();
    var doc = parser.parseFromString(icon, 'image/svg+xml');
    var rootElement = doc.documentElement;
    if (!rootElement) {
        return;
    }
    var width = parseInt(rootElement.getAttribute('width'), 10);
    var height = parseInt(rootElement.getAttribute('height'), 10);
    return { x: wRatio * width, y: hRatio * height };
}
