// graphql-output-generator/lib/graphql/create-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
/* this complexity is need for the test */
exports.dummyOutputType = new graphql_1.GraphQLObjectType({
    name: 'DummyQueryOutput',
    fields: {
        dummy: { type: graphql_1.GraphQLString },
        nested: {
            type: new graphql_1.GraphQLObjectType({
                name: 'nestedDummy',
                fields: {
                    nestedString: { type: graphql_1.GraphQLString }
                }
            })
        }
    }
});
exports.dummyQuery = {
    dummyQuery: {
        // resolve: dummyResolver,
        type: exports.dummyOutputType,
        args: {
            input: {
                type: new graphql_1.GraphQLInputObjectType({
                    name: 'DummyQueryInput',
                    fields: {
                        dummy: { type: graphql_1.GraphQLString }
                    }
                })
            }
        },
        description: 'Retrieve a Location'
    }
};
function createSchema(p) {
    var ret = {
        query: new graphql_1.GraphQLObjectType({
            name: 'Query',
            fields: p.queries || exports.dummyQuery
        })
    };
    if (p.mutations) {
        ret.mutation = new graphql_1.GraphQLObjectType({
            name: 'Mutations',
            fields: p.mutations
        });
    }
    return new graphql_1.GraphQLSchema(ret);
}
exports.createSchema = createSchema;
function createQuerySchema(queries) {
    return createSchema({ queries: queries });
}
exports.createQuerySchema = createQuerySchema;
function createMutationSchema(mutations) {
    return createSchema({ mutations: mutations });
}
exports.createMutationSchema = createMutationSchema;
