// @patternplate/client/lib/app/components/outside.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

class ClickOutside extends _react.Component {
  constructor(props) {
    super(props);
    this.handle = this.handle.bind(this);
    this.getContainer = this.getContainer.bind(this);
  }

  getContainer(ref) {
    this.container = ref;
  }

  render() {
    return _react.default.createElement("div", {
      className: this.props.className,
      ref: this.getContainer
    }, this.props.children);
  }

  componentDidMount() {
    global.document.addEventListener("click", this.handle, true);
  }

  componentWillUnmount() {
    global.document.removeEventListener("click", this.handle, true);
  }

  handle(e) {
    const props = this.props;
    const el = this.container;

    if (!el.contains(e.target)) {
      props.onClickOutside(e);
    }
  }

}

exports.default = ClickOutside;
