// @patternplate/client/lib/app/components/toggle-search.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _components = require("@patternplate/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = SearchButton;
exports.default = _default;

function SearchButton(props) {
  return _react.default.createElement(StyledLink, {
    debug: true,
    title: `Enable search ${props.shortcut.toString()}`,
    query: {
      "search-enabled": !props.enabled
    }
  }, _react.default.createElement(StyledIcon, {
    base: props.base,
    symbol: "search"
  }), "Search");
}

const StyledIcon = (0, _components.styled)(_components.Icon).withConfig({
  displayName: "toggle-search__StyledIcon"
})(["fill:", ";"], props => props.theme.colors.color);
const StyledLink = (0, _components.styled)(_components.Link).withConfig({
  displayName: "toggle-search__StyledLink"
})(["font-size:0;"]);
