// tslint-lerna-rules/lib/test.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint");
var ts = require("typescript");
var Test;
(function (Test) {
    var program;
    Test.RULES_DIRECTORY = 'lib';
    Test.FORMATTER_DIRECTORY = 'customFormatters/';
    Test.FILE_ENCODING = 'utf8';
    function runRule(opts) {
        var configuration = {
            extends: [],
            jsRules: new Map(),
            linterOptions: {},
            rules: new Map(),
            rulesDirectory: []
        };
        if (Array.isArray(opts.userOptions) && opts.userOptions.length > 0) {
            configuration.rules.set(opts.ruleName, {
                ruleName: opts.ruleName,
                ruleArguments: opts.userOptions
            });
        }
        else {
            configuration.rules.set(opts.ruleName, {
                ruleName: opts.ruleName
            });
        }
        var options = {
            formatter: 'json',
            fix: false,
            rulesDirectory: Test.RULES_DIRECTORY,
            formattersDirectory: Test.FORMATTER_DIRECTORY
        };
        if (opts.useTypeChecker) {
            var defaultHost_1 = ts.createCompilerHost({});
            program = ts.createProgram([opts.file.path], {}, __assign({}, defaultHost_1, { getSourceFile: function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i] = arguments[_i];
                    }
                    var path = args[0], version = args[1];
                    if (path === opts.file.path) {
                        return ts.createSourceFile(opts.file.path, opts.file.contents, version);
                    }
                    return defaultHost_1.getSourceFile.apply(defaultHost_1, args);
                } }));
        }
        var linter = new Lint.Linter(options, opts.useTypeChecker ? program : undefined);
        linter.lint(opts.file.path, opts.file.contents, configuration);
        return linter.getResult();
    }
    Test.runRule = runRule;
})(Test = exports.Test || (exports.Test = {}));
