// ts-jest/dist/postprocess.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var babel;
var istanbulPlugin;
var jestPreset;
function importBabelDeps() {
    if (babel) {
        return;
    }
    babel = require('babel-core');
    istanbulPlugin = require('babel-plugin-istanbul').default;
    jestPreset = require('babel-preset-jest');
}
var logger_1 = require("./logger");
function postProcessCode(compilerOptions, jestConfig, tsJestConfig, transformOptions, transpileOutput, filePath) {
    var postHook = exports.getPostProcessHook(compilerOptions, jestConfig, tsJestConfig);
    return postHook(transpileOutput, filePath, jestConfig, transformOptions);
}
exports.postProcessCode = postProcessCode;
function createBabelTransformer(options) {
    importBabelDeps();
    options = __assign({}, options, { plugins: options.plugins || [], presets: (options.presets || []).concat([jestPreset]) });
    delete options.cacheDirectory;
    delete options.filename;
    return function (codeSourcemapPair, filename, config, transformOptions) {
        var theseOptions = Object.assign({ filename: filename, inputSourceMap: codeSourcemapPair.map }, options);
        if (transformOptions && transformOptions.instrument) {
            theseOptions.auxiliaryCommentBefore = ' istanbul ignore next ';
            theseOptions.plugins = theseOptions.plugins.concat([
                [
                    istanbulPlugin,
                    {
                        cwd: config.rootDir,
                        exclude: [],
                    },
                ],
            ]);
        }
        return babel.transform(codeSourcemapPair.code, theseOptions);
    };
}
exports.getPostProcessHook = function (tsCompilerOptions, jestConfig, tsJestConfig) {
    if (tsJestConfig.skipBabel) {
        logger_1.logOnce('Not using any postprocess hook.');
        return function (input) { return input; };
    }
    var plugins = Array.from((tsJestConfig.babelConfig && tsJestConfig.babelConfig.plugins) || []);
    var babelOptions = __assign({}, tsJestConfig.babelConfig, { babelrc: tsJestConfig.useBabelrc || false, plugins: plugins, presets: tsJestConfig.babelConfig ? tsJestConfig.babelConfig.presets : [], sourceMaps: tsJestConfig.disableSourceMapSupport !== true });
    logger_1.logOnce('Using babel with options:', babelOptions);
    return createBabelTransformer(babelOptions);
};
