// @patternplate/client/lib/app/containers/navigation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationToolbar = exports.NavigationHeader = exports.default = void 0;

var _url = _interopRequireDefault(require("url"));

var _reactRedux = require("react-redux");

var _reactRouterRedux = require("react-router-redux");

var _redux = require("redux");

var _components = require("@patternplate/components");

var _actions = require("../actions");

var _navigation = _interopRequireDefault(require("../selectors/navigation"));

var _docs = _interopRequireDefault(require("../selectors/docs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = (0, _reactRedux.connect)(mapProps, mapDispatch)(_components.MainNavigation);

exports.default = _default;
const NavigationHeader = _components.MainNavigation.NavigationHeader;
exports.NavigationHeader = NavigationHeader;
const NavigationToolbar = _components.MainNavigation.NavigationToolbar;
exports.NavigationToolbar = NavigationToolbar;

function mapProps(state) {
  return {
    active: state.id,
    docs: (0, _docs.default)(state),
    componentsEnabled: state.componentsEnabled,
    navigation: (0, _navigation.default)(state),
    query: state.routing.locationBeforeTransitions.query
  };
}

function mapDispatch(dispatch) {
  return (0, _redux.bindActionCreators)({
    onItemClick(e) {
      e.preventDefault();

      if (!e.currentTarget.href) {
        return {
          type: "noop"
        };
      }

      const parsed = _url.default.parse(e.currentTarget.href);

      return (0, _reactRouterRedux.push)(`${parsed.pathname}?${parsed.query}`);
    },

    onLabelClick(query) {
      return (0, _actions.patchLocation)({
        query
      });
    }

  }, dispatch);
}
