// @patternplate/client/lib/app/containers/pattern.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactRedux = require("react-redux");

var _redux = require("redux");

var _reselect = require("reselect");

var demo = _interopRequireWildcard(require("../selectors/demo"));

var items = _interopRequireWildcard(require("../selectors/item"));

var _pattern = _interopRequireDefault(require("../components/pattern"));

var actions = _interopRequireWildcard(require("../actions"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class PatternContainer extends _react.default.Component {
  render() {
    const props = this.props;
    return _react.default.createElement(_pattern.default, {
      contents: props.contents,
      contentType: props.contentType,
      docs: props.docs,
      error: props.error,
      displayName: props.displayName,
      loading: props.loading,
      opacity: props.opacity,
      src: props.src,
      type: props.type,
      updated: props.updated,
      navigationEnabled: props.navigationEnabled,
      screenshot: props.screenshot
    });
  }

}

var _default = (0, _reactRedux.connect)(mapState, mapDispatch)(PatternContainer);

exports.default = _default;
const DEFAULT_PATTERN_CONTENTS = `
# :construction: Add documentation

> Undocumented software could not exist just as well.
>
> – The Voice of Common Sense

Currently there is no readme data for this pattern folder.
We left this friendly reminder for you to change that soon.

---

Help us to make this message more helpful on [GitHub](https://github.com/patternplate/patternplate).
`;
const NOT_FOUND = `
# Pattern not found

> Pretty sure this is not the component you are looking for.

We looked everywhere and could not find a single thing.

You might want to navigate back to [Home](/) or use the search.

---

Help us to make this message more helpful on [GitHub](https://github.com/patternplate/patternplate)
`;
const selectDocs = (0, _reselect.createSelector)(items.default, items.selectType, items.selectContents, (pattern, type, contents) => {
  if (pattern && pattern.type === 'folder') {
    return contents;
  }

  if (type === "not-found") {
    return NOT_FOUND;
  }

  return contents || DEFAULT_PATTERN_CONTENTS;
});

function mapState(state) {
  return {
    contents: state.demo.contents,
    displayName: items.selectName(state),
    docs: selectDocs(state),
    error: state.demo.error,
    loading: state.demo.fetching,
    opacity: state.opacity,
    src: demo.selectSrc(state),
    type: items.selectType(state),
    contentType: items.selectContentType(state),
    updated: state.demo.updated,
    navigationEnabled: state.navigationEnabled,
    screenshot: state.routing.locationBeforeTransitions.query.screenshot === "true"
  };
}

function mapDispatch(dispatch) {
  return (0, _redux.bindActionCreators)({
    onChange: actions.loadPatternDemo
  }, dispatch);
}
